Add `-Wnonunit-statement` to warn about discarded values in statement position

som-snytt opened the pull request:
## Summary

New compiler option `-Wnonunit-statement` warns about any interesting expression whose value is ignored because it is followed by another expression.

The new warnings are not included in `-Xlint`; they must be separately enabled with `-Wnonunit-statement`. Users of `-Wnonunit-statement` may also wish to enable a related option, `-Wvalue-discard`.

## Example

```scala
scala> :settings -Wnonunit-statement
scala> Some(3); "hello"
           ^
       warning: unused value
val res0: String = hello
```

## How to silence

The warning can be suppressed at a particular site by explicit ascription: `e: Unit`. This is the same convention used to ignore warnings about "value discard" conversions.

As always, any warning is also suppressible with `@nowarn`.

## Relation to other warnings

This generalizes the current warning for "pure" expressions, which only warns for limited cases where the expression can be proved not to be side-effecting.

`-Wvalue-discard` remains separate. It warns only when a value is discarded because it's the final expression in a block whose expected type is `Unit`, as [specified by the language](https://scala-lang.org/files/archive/spec/2.13/06-expressions.html#value-conversions).

## Details on what warns or doesn't

"Statement position" means statements in templates (i.e., constructors) and statements in blocks (everything in a block except the result expression). If a statement is an expression (and not a definition or assignment), then `-Wnonunit-statement` warns if the type of the expression is not `Unit`. There are heuristic exceptions, such as for methods that return `this.type`, where there is no loss of information. (The method is inherently side-effecting.) There is an additional flag whether to take an `if` statement as side-effecting, so that its result can be ignored.

By default, warns about `if` without `else`. Specify `-Wnonunit-if:false` to turn that off. (That also turns off `-Wvalue-discard` for those cases.)

The heuristic for `this.type`-returning methods intends to accommodate usual idioms updating unstable values.

## Motivation

As requested on Discord!

This is especially useful in pure-functional code where it's unusual for values ever to be intentionally discarded.

som-snytt commented:
Update to handle `Future(42): Unit` where the attachment is on the user expression, but we're linting the implicit `Future(42)(ec)` in `{ Future(42)(ec) ; () }`.

Typing the user expression yields the block, so additional work would be needed to copy attachments from the explicit apply to the implicit.

SethTisue commented:
I haven't seen so many yellow thumbs since my last all-day Simpsons-watching marathon. Looks like this will be popular if you can finish it.

som-snytt commented:
It's a bit niggling, but the warm reception means I can't stop at mile 18 despite the pain in my side.

som-snytt committed:
Add -Wnonunit-statement and -Wnonunit-if

Warn about any interesting value in statement position,
for some definition of interest.

Generalizes the current warning for "pure" expressions.

By default, warns about unibranch if, which lack an else
in user code.

Specify -Wnonunit-if:false to turn off warning about
unibranch if. (That also turns off -Wvalue-discard for
those cases.)

Improve explicit Unit check

Accommodate field when method is this.type

Handle Match

SethTisue commented:
@som-snytt anything you could use some feedback on or help with, here?

som-snytt commented:
I think it is just more slog.

som-snytt commented:
OK, I asked the internet brain. How many arms does a one-armed dog have if you call his tail an arm?

The answer is that if he's armed, do whatever he says and nod in agreement.

som-snytt commented:
It would be nice to provide an audit of `Unit` ascriptions that suppress a warning (or respectively suppress nothing). Maybe `-Y`.

Also a sibling option for `nowarn`. Otherwise, once you suppress, there is no opportunity to revisit. Grepping is harder than for `nowarn` usages.

I just needed that feature, A method was used a few times but only one warned because I'd previously slapped on the `Unit`.

What should the output look like? Just turning off suppression could result in many errors. Possibly specify a region of audit using `-opt:inline` syntax?

Or a general mechanism `-Yaudit-nowarn` where all suppressions are gathered, then reported by type and location.

som-snytt commented:
The improved term for elseless if is unibranching.

som-snytt commented:
The ongoing hassle of JDK8 support.
```
+nonunit-statement.scala:157: error: value of is not a member of object java.util.List
+    java.util.List.of[Int]()
```

som-snytt commented:
With improved partest, if I accidentally use API that is less ancient than the ancient JDK used for CI,
```
nonunit-statement.scala:159: error: value of is not a member of object java.util.List
    java.util.List.of[Int]()
                   ^
nonunit-statement.scala:54: warning: discarded non-Unit value
  Future(42): Unit        // nowarn { F(42)(ctx) }: Unit where annot is on F(42), TODO spurious -Wvalue-discard
        ^
1 warning
1 error
```

som-snytt commented:
To coin a phrase,
```
+error: -release is only supported on Java 9 and higher
```
... even if `-release 8` ...

som-snytt commented:
That makes the output of the two tests for bad options sensitive to the JVM version.
```
-error: bad options: -release 8 -Yresolve-term-conflict
+error: bad options: -Yresolve-term-conflict
```
I hope the `#partest` mechanism still works?

som-snytt commented:
This will require re-tweaks, as it still yields what I assume are spurious warnings on the standard library.
```
hashMapBuilder.addAll(xs)
```

som-snytt commented:
/rebuild

SethTisue commented:
@som-snytt I see CI is green. Is this ready for review? 2.13.9 is targeted for end-of-July, so the opportunity remains to get this in. I'm eager to help if I can.

SethTisue commented:
What's the design justification for making this separate from `-Wvalue-discard`?

SethTisue commented:
In the Scala 2 community build, we generally disable fatal warnings, and we don't really have any tools for asking "how does this change affect what warnings are generated", other than manual grepping and diffing. we should think a bit on whether to do some of the work in the community build contexts, or just find some volunteers to try out the PR validation snapshot on some suitable codebases.

som-snytt commented:
I'll remove the spurious commit and spend some time again. It looks like last time I still saw false positives.

"Value discard" is well-defined (where `{ stats ; expr }` becomes `{ stats ; expr ; () }`, but this warning is entirely a judgment about tree shapes or types and what the user might have intended.

What was the name of Johnny Carson's mind reader again? That's the name of this feature.

dwijnand commented on test/files/neg/nonunit-if.scala:
```diff
@@ -0,0 +1,172 @@
+class Absolution {
+  def f(i: Int): Int = i+1
+  import ExecutionContext.Implicits._
+  Future(42): Unit        // nowarn { F(42)(ctx) }: Unit where annot is on F(42), TODO spurious -Wvalue-discard
```
This TODO looks dead.  But it looks like there are spurious warnings on the uni-branched if's below (foreach/forall), which would be good to not emit ideally.

som-snytt commented:
I squashed for review without putting in more thought.

I see it needs a test around match expressions; in fact, I see that it should always check all case bodies, the way it always checks both branch of a conditional.

I'll try running it on this codebase again, and something typelevel, who are the target audience.

lrytz commented on test/files/neg/nonunit-if.check:
```diff
@@ -0,0 +1,57 @@
+nonunit-if.scala:62: warning: discarded non-Unit value
+      f(a)                  // warn, check is on
+       ^
+nonunit-if.scala:13: warning: unused value
```
should the message mention the `: Unit` workaround?

som-snytt replied:
I will follow up with that suggestion.

SethTisue replied:
@som-snytt I don't think you followed up on this?

SethTisue commented:
/rebuild

SethTisue commented:
sorry, but Jenkins isn't responding to `/rebuild`, so I think you'll need to rebase (or force-push some trivial change) to get Jenkins to trigger (now that the underlying disk space issue has been addressed)

som-snytt commented:
That's OK, I need to add the message that says, `See lrytz for a qwirkaround.`

Oh, I just discovered the correct spelling of `qwirkaround`.

som-snytt commented:
Maybe it's not Travis's fault after all.
```
[info] compiling 160 Scala sources and 3 Java sources to /home/travis/build/scala/scala/build/quick/classes/reflect ...

[warn] In the last 10 seconds, 5.051 (52.8%) were spent in GC. [Heap: 0.72GB free of 1.27GB, max 1.78GB] Consider increasing the JVM heap using `-Xmx` or try a different collector, e.g. `-XX:+UseG1GC`, for better performance.

[info] 2 optimizer info messages; re-run enabling -Wopt for details, or try -help

[info] done compiling

[info] compiling 327 Scala sources and 5 Java sources to /home/travis/build/scala/scala/build/quick/classes/compiler ...

[warn] In the last 9 seconds, 5.153 (57.4%) were spent in GC. [Heap: 0.74GB free of 1.30GB, max 1.78GB] Consider increasing the JVM heap using `-Xmx` or try a different collector, e.g. `-XX:+UseG1GC`, for better performance.

[warn] In the last 10 seconds, 7.284 (75.0%) were spent in GC. [Heap: 0.80GB free of 1.70GB, max 1.78GB] Consider increasing the JVM heap using `-Xmx` or try a different collector, e.g. `-XX:+UseG1GC`, for better performance.

[info] 101 optimizer info messages; re-run enabling -Wopt for details, or try -help
```

SethTisue commented:
Is this ready for review?

I think the PR description would benefit from a brief, canonical example of the kind of code this warns on.

som-snytt commented:
> Warn about any interesting value in statement position, for some definition of interest.

som-snytt commented:
Added a soupçon of rebase.

SethTisue commented:
> we should think a bit on whether to do some of the work in the community build contexts, or just find some volunteers to try out the PR validation snapshot on some suitable codebases.

Given that 2.13.9 is rapidly approaching, given the high value of this PR, and given that the new warnings must be explicitly enabled (they are not included in `-Xlint`), I think it's safe to ship this. We can always make adjustments in subsequent releases.

@som-snytt I revised the PR description pretty thoroughly; you want to check it over.

Can you take a stab at improving the message, as suggested by Lukas?

SethTisue commented:
@som-snytt I took this for a spin in two of my codebases and the only false positives I got were `StringBuilder` related and would presumably be fixed by #10047, so you might rebase to include that to aid testers.

SethTisue commented:
@som-snytt I think it would be helpful if the warning message said what type was discarded. I think this would be a big help for users to decide whether they want to fix the warning or ignore it. Including the relevant in the error message could also help with filtering; for example I might want to use `-Wconf:...` to ignore these warnings when they occur on a specific type such as `org.scalatest.Assertion`.

thought occasioned by trying this out in a third codebase I work on and getting:

```
[warn] /Users/tisue/lightbend-emoji/src/test/scala-2/com/lightbend/emoji/EmojiSpec.scala:55:35: unused value
[warn]       Emoji(e.codePoint).hashCode shouldBe (e.hashCode)
[warn]                                   ^
```

this is with ScalaTest and `import org.scalatest.matchers.should.Matchers._` — `shouldBe` returns an `org.scalatest.Assertion`. so that's an example of an API that will be troublesome in combination with this new flag

SethTisue commented:
It still nags at me that the distinction between `-Wvalue-discard` and this seems rather academic. How many actual users would want to enable `-Wvalue-discard` without also enabling this? In other words: shouldn't the functionality of `-Wnonunit-statement` simply be folded into `-Wvalue-discard`? (Perhaps with an opt-out if someone really wants to split this hair, comparable to the `-Wnonunit-if:false` opt-out you added.)

(Less importantly, I'm having doubts about whether we want to use the word "statement" in the flag name and documentation. And namingwise, I guess "statement" bothers me because it seems like importing terminology from other programming languages outside of the functional tradition. In Scala, a few things like `val` and `import` actually are statements... everything else is just expressions.)

som-snytt commented:
I would distinguish `-Wvalue-discard` only because it is well-defined and less prone to heuristic false positives.

They are statements when in statement position. The spec for "block" calls them statements, for example. Maybe time for a neologism. Blockpression. Exstatement.

I'll tear myself away from Fringe season 5 to update the message with the fine suggestions to suggest something.

som-snytt commented:
> Som day, Som day.

https://www.youtube.com/watch?v=UiF0FMs16dU

som-snytt commented:
Another idea is `-Ynag:Seth,tpolecat` for specialized nagging.

SethTisue commented:
> [youtube.com/watch?v=UiF0FMs16dU](https://www.youtube.com/watch?v=UiF0FMs16dU)

🎵 you gave me no warnin’ 🎵

lrytz commented on src/reflect/scala/reflect/internal/TreeInfo.scala:
```diff
@@ -351,6 +351,40 @@ abstract class TreeInfo {
+              receiver match {
+                case Apply(_, _) => Precedence(op.decoded).level == 0         // xs(i) += x
+                case _ => receiver.symbol != null &&
+                  (receiver.symbol.isGetter || receiver.symbol.isField)       // xs.addOne(x) for var xs
```
This part I don't understand... I don't see anything about `var` or `addOne`.

IIUC, it catches any method call on a field, no?

How about calls on local values / variables?

Also `addOne` and `+=` are defined with return type `this.type` - why do you need to handle them separately?

som-snytt replied:
If a value is not stable, in `v().addOne(x)`, I can witness different values in `v()` and the result. I think that is the gist of why, even if `addOne` is side-effecting and returns `this`, I've lost the value produced by `v`; I can't assume I can recover it by calling `v` again.

I see a test class `Variant`. I'll add more variants to ensure this check is not too wide a net. I think it applies only to `this`-typed methods. But is that even correct? I'm sure I noticed it by testing on this code base. But maybe `v.addOne(x)` is bad form.

lrytz replied:
👍  thanks. what i missed is that we're already narrowed down to a method with a this.type / singleton type here.

lrytz reviewed (commented):
Question: this warns - should it count as a `Unit` ascription?

```
class C {
  def f(x: Int): Int = 1 + x
  def g: Unit = f(0)
}
```

armanbilge commented:
I just took this for a spin on Skunk. Only two (spurious?) warnings!

https://github.com/tpolecat/skunk/blob/14c756d68ba88da4415699bea388149c43e4e962/modules/core/shared/src/main/scala-2/syntax/StringContextOps.scala#L79-L82

```
[error] /workspace/skunk/modules/core/shared/src/main/scala-2/syntax/StringContextOps.scala:79:36: unused value
[error]       val EncoderType      = typeOf[Encoder[_]]
[error]                                    ^
[error] /workspace/skunk/modules/core/shared/src/main/scala-2/syntax/StringContextOps.scala:81:36: unused value
[error]       val FragmentType     = typeOf[Fragment[_]]
[error]                                    ^
[error] two errors found
```

Seems like the issue is the `_`, if I change it to `Any` then the warning goes away.

som-snytt committed:
Tweak setInfo and do not warn on return of arg

som-snytt commented:
The Skunk warning is due to this call in the expansion. I added a commit that tweaks the signature, which may pass MiMA. There are probably other bits of reflect API that deserve similar tweaking.
```
$u.internal.reificationSupport.setInfo[$u.Symbol](symdef$EncoderType1, $u.NoType);
```

som-snytt commented:
@lrytz As to whether explicit result type means "I intend to discard on RHS", I think the answer is no, because `-Wvalue-discard` warns for this case.

Explicit result types are supplied often: to specify API, or to accommodate recursion.

I like the "transposed" form:
```
def g = f(42): Unit
```
which is somewhat equivalent but does show "intention to discard", and also is not redundant.

som-snytt committed:
Mention type of unused value and mitigation

lrytz commented on src/compiler/scala/tools/nsc/typechecker/RefChecks.scala:
```diff
@@ -1726,19 +1726,83 @@ abstract class RefChecks extends Transform {
+      // Parser adds suppressing attachment on `if (b) expr` when user has `-Wnonunit-if:false`.
+      def checkInterestingShapes(t: Tree): Boolean =
+        t match {
+          case If(_, thenpart, elsepart) => checkInterestingShapes(thenpart) | checkInterestingShapes(elsepart) // strict or
```
> strict or

why? i can't spot any side-effects

som-snytt replied:
I was afraid you would ask that. I remember needing to descend both branches. I'll see what test breaks.

I also remember that the match doesn't visit all branches, which is not consistent.

som-snytt replied:
Well, the test passed, so I can only plead "for historical reasons". I'll add a VBAR.

som-snytt committed:
Add VBAR per eagle-eyed review

som-snytt commented:
Since I have to rebase anyway, sigh, I'll add a code comment where lrytz had to read it twice. I had to read it more than that.

SethTisue commented:
needs checkfile update, looks like

som-snytt commented:
because the other warning jumped the queue, need to update on the first commit
```
!!    1 - neg/nonunit-if.scala                      [output differs]
% diff /home/jenkins/workspace/scala-2.13.x-validate-main/test/files/neg/nonunit-if.check /home/jenkins/workspace/scala-2.13.x-validate-main/test/files/neg/nonunit-if-neg.log
@@ -1,7 +1,7 @@
-nonunit-if.scala:58: warning: discarded non-Unit value
+nonunit-if.scala:58: warning: discarded non-Unit value of type U
     if (!isEmpty) f(a)      // warn, check is on
                    ^
-nonunit-if.scala:62: warning: discarded non-Unit value
+nonunit-if.scala:62: warning: discarded non-Unit value of type Boolean
       f(a)                  // warn, check is on
        ^
 nonunit-if.scala:13: warning: unused value
```

SethTisue commented:
🎉

sideeffffect commented:
Is there a chance that we will have something like this in Dotty?

SethTisue commented:
@sideeffffect thread is https://contributors.scala-lang.org/t/replacement-for-xlint-in-scala3/5241
